### PR TITLE
[Core] Chunked Prefill support for Multi Step Scheduling

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
Add Chunked Prefill support for Multi Step Scheduling. 

Based on, https://github.com/vllm-project/vllm/issues/7528, this PR adds both the `Force Single Step` and `Ignore Prefill` policies.

The `Ignore Prefill` policy implemented as the default policy with a `VLLM_MULTI_STEP_CHUNKED_PREFILL_SINGLE_STEP_POLICY` to switch to the `Force Single Step` policy. 

Benchmark : 

 Please find A100 benchmark results [here](https://docs.google.com/spreadsheets/d/1O795J0_IRTkYxswCJ4_fycLlmMIu9EyIL9eCdnqCU3M/edit?usp=sharing) 
---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/core/scheduler.py [hunk 1]
```diff
@@ -1202,7 +1217,8 @@ def _append_slots(
                 the new source and destination block indices for the appended
                 slots.
         """
-        num_lookahead_slots = self._get_num_lookahead_slots(is_prefill=False)
+        num_lookahead_slots = self._get_num_lookahead_slots(
+            is_prefill=seq_group.is_prefill())
```
Comment (SolitaryThinker) [inline vllm/core/scheduler.py:1221]: Spent some time looking into PP. The assertion issue may be related to this change. Something is causing the batch to generate logits for the prefills in the batch and since `len(sample_indices) == 0) as they don't perform sampling, the assertion fails on `assert logits_applied == logits.shape[0]` in `_apply_min_tokens_penalty` since logits_applied is the sum of all `sampled_indices` in the batch

Comment (varun-sundar-rabindranath) [reply vllm/core/scheduler.py:1221]: I see. Thanks @SolitaryThinker. I believe it should be 
```
        num_lookahead_slots = self._get_num_lookahead_slots(is_prefill=seq_group.is_prefill() and seq_group.do_sample)
```
instead of just 
```
        num_lookahead_slots = self._get_num_lookahead_slots(is_prefill=False)
```
I haven't been able to reproduce this assertion yet. I'll keep at it. When you find some time, can you try this out as well ? Thanks.

File: vllm/engine/async_llm_engine.py [hunk 2]
```diff
@@ -346,7 +347,8 @@ async def step_async(
         # Finish the current step for all the sequence groups.
         if self.scheduler_config.is_multi_step:
             for seq_group in seq_group_metadata_list:
-                seq_group.finish_step()
+                if seq_group.state.remaining_steps > 0:
+                    seq_group.finish_step()
```
Comment (comaniac) [inline vllm/engine/async_llm_engine.py:351]: Could you comment on this?

File: tests/multi_step/test_correctness.py [hunk 3]
```diff
@@ -23,17 +48,36 @@
 ]
 
 
-async def completions_with_server_args(prompts: List[str], model_name: str,
-                                       server_cli_args: List[str]):
+class EnvContextManager():
```
Comment (comaniac) [inline tests/multi_step/test_correctness.py:51]: Can you look around the tests and see if we already have a similar utility? If so please reuse it; otherwise please move this to the right place so that other tests in the future can use it.

File: vllm/attention/backends/flash_attn.py [hunk 4]
```diff
@@ -300,8 +301,60 @@ def decode_metadata(self) -> Optional["FlashAttentionMetadata"]:
             block_tables=self.block_tables[self.num_prefills:],
             use_cuda_graph=self.use_cuda_graph,
         )
+
         return self._cached_decode_metadata
 
+    # TODO (varun) : Try using decode_metadata here. We hit some asserts in
+    # advance_step - but that seems resolvable.
```
Comment (comaniac) [inline vllm/attention/backends/flash_attn.py:308]: Can you elaborate more on this? Specifically how it can be solved and what's the plan?

**Review by comaniac — COMMENTED** [review]
Thanks for the great work! Leave some comments.

File: tests/multi_step/test_correctness.py [hunk 5]
```diff
@@ -23,17 +48,36 @@
...
+        os.environ.clear()
+        os.environ.update(self.os_env)
+
+
+async def completions_with_server_args(prompts: List[str],
+                                       model_name: str,
+                                       server_cli_args: List[str],
+                                       with_env: dict = {}):  # noqa: B006
+    # env setup
+    os.environ.update(with_env)
```
Comment (varun-sundar-rabindranath) [inline tests/multi_step/test_correctness.py:70]: TODO : fix stray update.

File: tests/multi_step/test_correctness.py [hunk 6]
```diff
@@ -23,17 +48,36 @@
...
+
+    def __enter__(self):
+        os.environ.update(self.add_env)
+
+    def __exit__(self, *args, **kwargs):
+        os.environ.clear()
+        os.environ.update(self.os_env)
+
+
+async def completions_with_server_args(prompts: List[str],
```
[the remark this answers is not in the document]
Comment (comaniac) [reply tests/multi_step/test_correctness.py:65]: Oops yes sorry for the typo.

File: vllm/core/scheduler.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply vllm/core/scheduler.py:1221]: https://github.com/vllm-project/vllm/pull/7814/commits/976d03275dd7a5f0d125769484dbe6a42a264c64 fixes the PP issue. The SamplingMetadata objects were being clobbered due to the SamplingMetadataCache reset.

**Comment by varun-sundar-rabindranath** [comment]
Bad performance - just `multi-step` is better than `multi-step + chunked-prefill` closing this in favor of https://github.com/vllm-project/vllm/pull/8378

**Comment by varun-sundar-rabindranath** [comment]
@tlrmchlsmth @comaniac @SolitaryThinker I'll handle the relevant comments in the new PR
